devito/ir/equations/algorithms.py

```diff
@@ -162,74 +162,124 @@ def concretize_subdims(exprs, **kwargs):
+    thickness = tuple((v, d._thickness_map[k]) for k, v in tkns_subs.items())
 
-@_concretize_subdims.register(ConditionalDimension)
-def _(d, mapper, sregistry):
-    # TODO: to be implemented as soon as we drop the counter machinery in
-    # Grid.__subdomain_finalize__
-    # TODO: call `_concretize_subdims(d.parent, mapper)` as the parent might be
-    # a SubDimension!
-    pass
+    mapper[d] = d._rebuild(symbolic_min=left, symbolic_max=right, thickness=thickness)
 
 
-@_concretize_subdims.register(MultiSubDimension)
-def _(d, mapper, sregistry):
-    if not d.is_abstract:
-        # TODO: for now Grid.__subdomain_finalize__ creates the thickness, but
-        # soon it will be done here instead
+@_concretize_subdims.register(ConditionalDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
+        # Already have a substitution for this dimension
         return
 
-    pd = d.parent
+    _concretize_subdims(d.parent, mapper, rebuilt, sregistry)
+
+    kwargs = {}
+
+    # Parent may be a subdimension
+    if d.parent in mapper:
+        kwargs['parent'] = mapper[d.parent]
+
+    # Condition may contain subdimensions
+    if d.condition is not None:
+        for v in d.condition.free_symbols:
+            _concretize_subdims(v, mapper, rebuilt, sregistry)
 
-    subs = mapper[pd]
+        if any(v in mapper for v in d.condition.free_symbols):
+            # Substitute into condition
+            kwargs['condition'] = d.condition.subs(mapper)
 
-    if d in subs:
+    if kwargs:
+        # Rebuild if parent or condition need replacing
+        mapper[d] = d._rebuild(**kwargs)
+
+
+@_concretize_subdims.register(MultiSubDimension)
+def _(d, mapper, rebuilt, sregistry):
+    if d in mapper:
         # Already have a substitution for this dimension
         return
 
-    name = sregistry.make_name(prefix=d.name)
-    ltkn, rtkn = MultiSubDimension._symbolic_thickness(name)
+    tkns0 = MultiSubDimension._symbolic_thickness(d.name)
+    tkns1 = [tkn._rebuild(name=sregistry.make_name(prefix=tkn.name))
+             for tkn in tkns0]
 
-    thickness = (ltkn, rtkn)
+    kwargs = {'thickness': tuple(tkns1), 'functions': d.functions}
 
-    subs[d] = d._rebuild(d.name, pd, thickness)
+    idim0 = d.implicit_dimension
+    if idim0 is not None:
+        try:
+            # Get a preexisiting substitution if one exists
+            idim1 = rebuilt[idim0]
+            # If a substitution exists for the implicit dimension,
+            # then there is also one for the function
+            functions = rebuilt[d.functions]
+        except KeyError:
+            iname = sregistry.make_name(prefix=idim0.name)
+            rebuilt[idim0] = idim1 = idim0._rebuild(name=iname)
+
+            fdims = (idim1,) + (d.functions.dimensions[1:])
+            frebuilt = d.functions._rebuild(dimensions=fdims, function=None,
+                                            halo=None, padding=None,
```

Comment:
are halo and padding really necessary here

Reply:
Yeah, I looked into removing them, but the problem is, if you set `function=None` (which is necessary here), there is no way of checking the dimensions have been changed during a rebuild, since you can't check against the `function` kwarg. You need to specify `halo=None` and `padding=None` when changing the dimensions, as a bunch of mappers need reconstructing with the new dimensions. It throws an error otherwise.